lib/mayaUsd/ufe/UsdTransform3dMayaXformStack.cpp

```diff
@@ -0,0 +1,708 @@
+        const Ufe::Path&      path,
+        const UsdGeomXformOp& op,
+        const UsdTimeCode&    writeTime_
+    ) : Ufe::SetVector3dUndoableCommand(path), fOp(op),
+        // Always read from proxy shape time.
+        fReadTime(getTime(path)),
+        fWriteTime(writeTime_),
+        fPrevOpValue(getValue(op.GetAttr(), readTime())),
+        fNewOpValue(newOpValue),
+        fAttrName(op.GetOpName())
+    {}
+
+    // Have separate execute override which is value-only, as default execute()
+    // calls redo, and undo / redo will eventually not be value only and will
+    // support propertySpec / primSpec management.
+    void execute() override { setValue(fNewOpValue); }
+
+    void recreateOp() {
+        auto sceneItem = std::dynamic_pointer_cast<UsdSceneItem>(
+            Ufe::Hierarchy::createItem(path())); TF_AXIOM(sceneItem);
+        auto prim = sceneItem->prim(); TF_AXIOM(prim);
+        auto attr = prim.GetAttribute(fAttrName); TF_AXIOM(attr);
+        fOp       = UsdGeomXformOp(attr);
+    }
+
+    void undo() override {
+        // Re-create the XformOp, as the underlying prim may have been
+        // invalidated by later commands.
+        recreateOp();
+        setValue(fPrevOpValue);
+    }
+    void redo() override {
+        // Re-create the XformOp, as the underlying prim may have been
+        // invalidated by earlier commands.
+        recreateOp();
+        setValue(fNewOpValue);
+    }
+
+    void setValue(const VtValue& v) { fOp.GetAttr().Set(v, fWriteTime); }
+
+    UsdTimeCode readTime() const { return fReadTime; }
+    UsdTimeCode writeTime() const { return fWriteTime; }
+
+protected:
+    VtValue fNewOpValue;
+};
+
+// UsdRotatePivotTranslateUndoableCmd uses hard-coded USD common transform API
+// single pivot attribute name, not reusable.
+template<class V>
+class UsdVecOpUndoableCmd : public UsdTRSUndoableCmdBase {
+public:
+    UsdVecOpUndoableCmd(
+        const V&              v,
+        const Ufe::Path&      path,
+        const UsdGeomXformOp& op,
+        const UsdTimeCode&    writeTime
+    ) : UsdTRSUndoableCmdBase(VtValue(v), path, op, writeTime)
+    {}
+
+    // Executes the command by setting the translation onto the transform op.
+    bool set(double x, double y, double z) override
+    {
+        fNewOpValue = V(x, y, z);
+
+        setValue(fNewOpValue);
+        return true;
+    }
+};
+
+class UsdRotateOpUndoableCmd : public UsdTRSUndoableCmdBase {
+public:
+    UsdRotateOpUndoableCmd(
+        const GfVec3f&        r,
+        const Ufe::Path&      path,
+        const UsdGeomXformOp& op,
+        CvtRotXYZToAttrFn     cvt,
+        const UsdTimeCode&    writeTime
+    ) : UsdTRSUndoableCmdBase(VtValue(r), path, op, writeTime), 
+        _cvtRotXYZToAttr(cvt)
+    {}
+
+    // Executes the command by setting the rotation onto the transform op.
+    bool set(double x, double y, double z) override
+    {
+        fNewOpValue = _cvtRotXYZToAttr(x, y, z);
```

Comment:
Rotation is special, as it needs a conversion function from UFE's RotXYZ to whatever the attribute is storing.